File: sidecar/agent/taskstore.py
"""Minimal task and event persistence — stores user-facing tasks and events under ~/.zwork/tasks.json."""

from __future__ import annotations

import json
import os
import tempfile
from dataclasses import dataclass, asdict
from pathlib import Path
from typing import Any

from .home import zwork_home
from .utils import now_ms, uid


@dataclass
class Task:
    id: str
    title: str
    column: str  # "inbox" | "todo" | "doing" | "done"
    created_at: int
    updated_at: int
    due_date: str | None = None  # "YYYY-MM-DD"
    completed_at: int | None = None
    description: str = ""
    assignee: str = ""  # "me" | "zwork" | ""
    priority: str = "medium"  # "low" | "medium" | "high"


@dataclass
class CalendarEvent:
    id: str
    title: str
    date: str  # "YYYY-MM-DD"
    created_at: int
    start_time: str | None = None  # "HH:MM"
    end_time: str | None = None  # "HH:MM"


def _path() -> Path:
    """Return the JSONL storage path for the given *chat_id*."""
    return zwork_home() / "tasks.json"


def _load_data() -> dict[str, list[dict[str, Any]]]:
    p = _path()
    if not p.exists():
        return {"tasks": [], "events": []}
    try:
        content = p.read_text(encoding="utf-8")
        data = json.loads(content)
        if not isinstance(data, dict):
            return {"tasks": [], "events": []}
        return {
            "tasks": data.get("tasks") or [],
            "events": data.get("events") or [],
        }
    except Exception:
        return {"tasks": [], "events": []}


def _save_data(data: dict[str, list[dict[str, Any]]]) -> None:
    p = _path()
    formatted = json.dumps(data, indent=2)
    fd, tmp = tempfile.mkstemp(dir=p.parent, suffix=".tmp")
    try:
        os.write(fd, formatted.encode("utf-8"))
        os.close(fd)
        fd = -1
        os.replace(tmp, p)
    except BaseException:
        if fd >= 0:
            os.close(fd)
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise


# --- Task CRUD Operations ---


def get_tasks() -> list[Task]:
    data = _load_data()
    out = []
    for t in data.get("tasks", []):
        try:
            out.append(
                Task(
                    id=t["id"],
                    title=t["title"],
                    column=t["column"],
                    created_at=t["created_at"],
                    updated_at=t["updated_at"],
                    due_date=t.get("due_date"),
                    completed_at=t.get("completed_at"),
                    description=t.get("description", ""),
                    assignee=t.get("assignee", ""),
                    priority=t.get("priority", "medium"),
                )
            )
        except (KeyError, TypeError):
            continue
    return out


def save_task(
    title: str,
    column: str = "inbox",
    due_date: str | None = None,
    task_id: str | None = None,
    description: str = "",
    assignee: str = "",
    priority: str = "medium",
) -> Task:
    data = _load_data()
    tasks_list = data["tasks"]
    now = now_ms()

    # Find existing or create new
    existing_idx = -1
    if task_id:
        for i, t in enumerate(tasks_list):
            if t["id"] == task_id:
                existing_idx = i
                break

    if existing_idx >= 0:
        t_data = tasks_list[existing_idx]
        prev_column = t_data.get("column", "inbox")
        completed_at = t_data.get("completed_at")

        # If transitioning to "done", set completed_at
        if column == "done" and prev_column != "done":
            completed_at = now
        elif column != "done":
            completed_at = None

        t_data.update(
            {
                "title": title,
                "column": column,
                "due_date": due_date,
                "completed_at": completed_at,
                "updated_at": now,
                "description": description or t_data.get("description", ""),
                "assignee": assignee or t_data.get("assignee", ""),
                "priority": priority or t_data.get("priority", "medium"),
            }
        )
        task = Task(**t_data)
    else:
        # Create a new task
        completed_at = now if column == "done" else None
        task = Task(
            id=uid(),
            title=title,
            column=column,
            created_at=now,
            updated_at=now,
            due_date=due_date,
            completed_at=completed_at,
            description=description,
            assignee=assignee,
            priority=priority,
        )
        tasks_list.append(asdict(task))

    _save_data(data)
    return task


def update_task_column(task_id: str, column: str) -> Task | None:
    data = _load_data()
    tasks_list = data["tasks"]
    now = now_ms()

    for t_data in tasks_list:
        if t_data["id"] == task_id:
            prev_column = t_data.get("column", "inbox")
            completed_at = t_data.get("completed_at")

            if column == "done" and prev_column != "done":
                completed_at = now
            elif column != "done":
                completed_at = None

            t_data.update(
                {
                    "column": column,
                    "completed_at": completed_at,
                    "updated_at": now,
                }
            )
            task = Task(**t_data)
            _save_data(data)
            return task
    return None


def delete_task(task_id: str) -> bool:
    """Remove the task record file for *task_id*."""
    data = _load_data()
    tasks_list = data["tasks"]
    filtered = [t for t in tasks_list if t["id"] != task_id]
    if len(filtered) == len(tasks_list):
        return False
    data["tasks"] = filtered
    _save_data(data)
    return True


# --- Calendar Event CRUD Operations ---


def get_events() -> list[CalendarEvent]:
    data = _load_data()
    out = []
    for e in data.get("events", []):
        try:
            out.append(
                CalendarEvent(
                    id=e["id"],
                    title=e["title"],
                    date=e["date"],
                    created_at=e["created_at"],
                    start_time=e.get("start_time"),
                    end_time=e.get("end_time"),
                )
            )
        except (KeyError, TypeError):
            continue
    return out


def save_event(
    title: str,
    date: str,
    start_time: str | None = None,
    end_time: str | None = None,
    event_id: str | None = None,
) -> CalendarEvent:
    data = _load_data()
    events_list = data["events"]
    now = now_ms()

    # Find existing or create new
    existing_idx = -1
    if event_id:
        for i, e in enumerate(events_list):
            if e["id"] == event_id:
                existing_idx = i
                break

    if existing_idx >= 0:
        e_data = events_list[existing_idx]
        e_data.update(
            {
                "title": title,
                "date": date,
                "start_time": start_time,
                "end_time": end_time,
            }
        )
        event = CalendarEvent(**e_data)
    else:
        event = CalendarEvent(
            id=uid(),
            title=title,
            date=date,
            created_at=now,
            start_time=start_time,
            end_time=end_time,
        )
        events_list.append(asdict(event))

    _save_data(data)
    return event


def delete_event(event_id: str) -> bool:
    data = _load_data()
    events_list = data["events"]
    filtered = [e for e in events_list if e["id"] != event_id]
    if len(filtered) == len(events_list):
        return False
    data["events"] = filtered
    _save_data(data)
    return True

File: sidecar-rust/src/taskstore.rs
use serde::{Deserialize, Serialize};
use serde_json::{json, Value};
use std::fs;
use std::path::PathBuf;
use std::time::{SystemTime, UNIX_EPOCH};
use uuid::Uuid;

use crate::paths::tasks_path;

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as u64)
        .unwrap_or(0)
}

fn uid() -> String {
    Uuid::new_v4().simple().to_string()
}

// ─── Data Structures ──────────────────────────────────────────────────────────

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Task {
    pub id: String,
    pub title: String,
    #[serde(default = "default_inbox")]
    pub column: String,
    pub created_at: u64,
    pub updated_at: u64,
    #[serde(default)]
    pub due_date: Option<String>,
    #[serde(default)]
    pub completed_at: Option<u64>,
    #[serde(default)]
    pub description: String,
    #[serde(default)]
    pub assignee: String,
    #[serde(default = "default_medium")]
    pub priority: String,
}

fn default_inbox() -> String { "inbox".to_string() }
fn default_medium() -> String { "medium".to_string() }

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct CalendarEvent {
    pub id: String,
    pub title: String,
    pub date: String,
    pub created_at: u64,
    #[serde(default)]
    pub start_time: Option<String>,
    #[serde(default)]
    pub end_time: Option<String>,
}

#[derive(Serialize, Deserialize, Default)]
struct TaskStoreData {
    #[serde(default)]
    tasks: Vec<Task>,
    #[serde(default)]
    events: Vec<CalendarEvent>,
}

// ─── Persistence ──────────────────────────────────────────────────────────────

fn load_data() -> TaskStoreData {
    let p = tasks_path();
    if !p.exists() {
        return TaskStoreData::default();
    }
    let content = match fs::read_to_string(&p) {
        Ok(c) => c,
        Err(_) => return TaskStoreData::default(),
    };
    serde_json::from_str(&content).unwrap_or_default()
}

fn save_data(data: &TaskStoreData) {
    let p = tasks_path();
    if let Ok(content) = serde_json::to_string_pretty(data) {
        let tmp = p.with_extension("tmp");
        if fs::write(&tmp, content).is_ok() {
            let _ = fs::rename(tmp, p);
        }
    }
}

// ─── Task CRUD ────────────────────────────────────────────────────────────────

pub fn get_tasks() -> Vec<Task> {
    load_data().tasks
}

pub fn create_task(
    title: String,
    column: Option<String>,
    description: Option<String>,
    priority: Option<String>,
    due_date: Option<String>,
    assignee: Option<String>,
) -> Task {
    let mut data = load_data();
    let now = now_ms();
    let task = Task {
        id: uid(),
        title,
        column: column.unwrap_or_else(default_inbox),
        created_at: now,
        updated_at: now,
        due_date,
        completed_at: None,
        description: description.unwrap_or_default(),
        assignee: assignee.unwrap_or_default(),
        priority: priority.unwrap_or_else(default_medium),
    };
    data.tasks.push(task.clone());
    save_data(&data);
    task
}

pub fn update_task(
    task_id: &str,
    title: Option<String>,
    column: Option<String>,
    description: Option<String>,
    priority: Option<String>,
    due_date: Option<String>,
    assignee: Option<String>,
) -> Option<Task> {
    let mut data = load_data();
    let task = data.tasks.iter_mut().find(|t| t.id == task_id)?;
    if let Some(t) = title { task.title = t; }
    if let Some(c) = column {
        // Auto-set completed_at when moving to done
        if c == "done" && task.column != "done" {
            task.completed_at = Some(now_ms());
        }
        task.column = c;
    }
    if let Some(d) = description { task.description = d; }
    if let Some(p) = priority { task.priority = p; }
    if due_date.is_some() { task.due_date = due_date; }
    if let Some(a) = assignee { task.assignee = a; }
    task.updated_at = now_ms();
    let result = task.clone();
    save_data(&data);
    Some(result)
}

pub fn update_task_column(task_id: &str, column: &str) -> Option<Task> {
    let mut data = load_data();
    let task = data.tasks.iter_mut().find(|t| t.id == task_id)?;
    if column == "done" && task.column != "done" {
        task.completed_at = Some(now_ms());
    }
    task.column = column.to_string();
    task.updated_at = now_ms();
    let result = task.clone();
    save_data(&data);
    Some(result)
}

pub fn delete_task(task_id: &str) -> bool {
    let mut data = load_data();
    let before = data.tasks.len();
    data.tasks.retain(|t| t.id != task_id);
    if data.tasks.len() != before {
        save_data(&data);
        true
    } else {
        false
    }
}

// ─── Event CRUD ───────────────────────────────────────────────────────────────

pub fn get_events() -> Vec<CalendarEvent> {
    load_data().events
}

pub fn create_event(
    title: String,
    date: String,
    start_time: Option<String>,
    end_time: Option<String>,
) -> CalendarEvent {
    let mut data = load_data();
    let event = CalendarEvent {
        id: uid(),
        title,
        date,
        created_at: now_ms(),
        start_time,
        end_time,
    };
    data.events.push(event.clone());
    save_data(&data);
    event
}

pub fn delete_event(event_id: &str) -> bool {
    let mut data = load_data();
    let before = data.events.len();
    data.events.retain(|e| e.id != event_id);
    if data.events.len() != before {
        save_data(&data);
        true
    } else {
        false
    }
}
